### ISSUE TITLE
Support for importing/exporting semaphores

### ISSUE THREAD

**Issue by kpet** [google/iree#3378]
Vulkan enables applications to import external semaphores or export semaphores via [VK_KHR_external_semaphore_fd](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_KHR_external_semaphore_fd.html). This enables low-cost synchronization with external entities (either other APIs or other devices). Supporting this would greatly improve the overall performance when IREE is integrated into a multi-device pipeline.

**Comment by antiagainst** [comment]
Thanks @kpet! Yes, this is certainly something on our roadmap. Being able to interact with the embedding application in a non-intrusive way is one of the key goals IREE tries to hit. Basically the application should be able to bring its own device/queue/whatever and IREE will use those when instructed. We have APIs like [`iree_hal_vulkan_driver_wrap_device`](https://github.com/google/iree/blob/781606592942e6609bf2ffd1244aad2d7c7648d4/iree/hal/vulkan/api.h#L188-L218) right now and the rest needs to be built out, including using external semaphores.

**Comment by benvanik** [comment]
Yep! Same with import/export of memory; these features will likely start to land early next year as we need to continue building out the compiler support for async behaviors and can co-design the cross-platform APIs with their usage in IREE-generated modules.

**Comment by benvanik** [comment]
One bit of prework would be understanding how external semaphores work with timeline semaphores - that's still unclear to me. Would be worth prototyping something to ensure we can smoothly integrate things.

**Comment by benvanik** [comment]
The concrete work here is now adding APIs that use `iree_wait_handle_t` to import/export fds/HANDLEs/etc.
